runtime/src/accounts_background_service.rs

```diff
@@ -0,0 +1,160 @@
+};
+use crossbeam_channel::{Receiver, Sender};
+use log::*;
+use solana_measure::measure::Measure;
+use solana_sdk::clock::Slot;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, RwLock,
+};
+use std::thread::{self, sleep, Builder, JoinHandle};
+use std::time::Duration;
+
+const INTERVAL_MS: u64 = 100;
+const SHRUNKEN_ACCOUNT_PER_SEC: usize = 250;
+const SHRUNKEN_ACCOUNT_PER_INTERVAL: usize =
+    SHRUNKEN_ACCOUNT_PER_SEC / (1000 / INTERVAL_MS as usize);
+
+pub type SnapshotRequestSender = Sender<SnapshotRequest>;
+pub type SnapshotRequestReceiver = Receiver<SnapshotRequest>;
+
+pub struct SnapshotRequest {
+    pub snapshot_root_bank: Arc<Bank>,
+    pub status_cache_slot_deltas: Vec<BankSlotDelta>,
+}
+
+pub struct SnapshotItems {
+    pub snapshot_config: SnapshotConfig,
+    pub snapshot_request_receiver: SnapshotRequestReceiver,
+    pub accounts_package_sender: AccountsPackageSender,
+}
+
+pub struct AccountsBackgroundService {
+    t_background: JoinHandle<()>,
+}
+
+impl AccountsBackgroundService {
+    pub fn new(
+        bank_forks: Arc<RwLock<BankForks>>,
+        exit: &Arc<AtomicBool>,
+        snapshot_items: Option<SnapshotItems>,
+    ) -> Self {
+        info!("AccountsBackgroundService active");
+        let exit = exit.clone();
+        let mut consumed_budget = 0;
+        let mut last_cleaned_slot = 0;
+        let t_background = Builder::new()
+            .name("solana-accounts-background".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                // Grab the current root bank
+                let bank = bank_forks.read().unwrap().root_bank().clone();
+
+                // Check to see if there were any requests for snapshotting banks
+                // < the current root bank `bank` above.
+
+                // Claim: Any snapshot request for slot `N` found here means that cleanup has
+                // been called at most on slot `N`
+                //
+                // Proof: Assume for contradiction that we find a snapshot request for slot `N` here,
+                // but cleanup has already happened on some slot `M > N`.
+                //
+                // Then that means on some previous iteration of this loop, we found `bank_forks.root_bank() == M`,
+                // but did not see the snapshot for `N`. However, this is impossible because BankForks.set_root()
+                // will always flush the snapshot request for `N` to the snapshot request channel
+                // before setting a root `M > N`.
+                let snapshotted_slot = snapshot_items.as_ref().and_then(|snapshot_items| {
+                    let SnapshotItems {
+                        snapshot_config,
+                        snapshot_request_receiver,
+                        accounts_package_sender,
+                    } = snapshot_items;
+                    Self::process_snapshot_requests(
```

Comment:
how about moving this method to SnapshotItems or (SnapshotRequestHandler), so the code would nicely read as hear:

```rust
let snapshotted_slot = snapshot_request_handler.map(|handler| handler.handle_snapshot_request());
```

Reply:
or you can stick with `process_`/`Processor`. :)

Reply:
good idea!